Async migrations 0002 migration clarify wording (#2868)
PostHog/posthog.com: contents/docs/self-host/configure/async-migrations/0002-events-sample-by.md

1. **Performance:** The new schema leads to a performance (speed) improvement for various PostHog queries
2. **Upgrading ClickHouse:** Changing the schema was necessary to unblock the upgrading of ClickHouse, which is something we aim to complete in PostHog version 1.33.0 and can also bring massive performance improvements.

> Note: During the migration events ingestion from Kafka to ClickHouse will be paused for a brief period. There won't be any data loss as we'll be consuming all the events from Kafka later. However during that brief period you might see inconsistencies in the data in the PostHog app.

## Preparation

> 1. Make sure you have enough free space in your ClickHouse instance (we also run this as a `precheck` and `healthcheck`)

Reviewer: ```suggestion
1. Make sure you have enough free space in your ClickHouse instance. We certify this via a preflight check before running the migration, but it is good that you're also aware of the requirement.
```

> 2. Make sure there's enough disk space in Kafka and verify that the retention policies are reasonable (ClickHouse events ingestion will be paused during the migration and to make sure we don't lose any data we'll want to make sure they won't expire too fast from Kafka).

Reviewer: ```suggestion
2. Make sure there's enough disk space in Kafka and verify that the retention policies are reasonable (ClickHouse event ingestion will be paused during the migration and to make sure we don't lose any data we'll want to make sure they won't expire too fast from Kafka).
```

## Operations

1. Create a new table with the updated schema: `SAMPLE BY cityHash64(distinct_id)` + `ORDER BY (team_id, toDate(timestamp), event, cityHash64(distinct_id), cityHash64(uuid))`
2. Start backfilling the new table (online) with data from older partitions

> 3. Detach the `events_mv` materialized view so we stop ingesting events from Kafka to ClickHouse, this makes sure we don't lose any events during the table rename and the events will all be waiting in Kafka. From this point until step 7 we might see inconsistencies in the app.

Reviewer: ```suggestion
3. Detach the `events_mv` materialized view so we stop ingesting events from Kafka to ClickHouse. This makes sure we don't lose any events while switching the tables and all the events will be waiting in Kafka. From this point until step 7 we might not see new events in the PostHog UI.
```

### Will this migration cause any data loss?

> No. During the migration events ingestion from Kafka to ClickHouse will be paused for a brief period. There won't be any data loss as we'll be consuming all the events from Kafka later. Furthermore this migration duplicates the events table and keeps the old table as a backup so we can always restore it.

Reviewer: ```suggestion
No. During the migration event ingestion from Kafka to ClickHouse will be paused for a brief period. There won't be any data loss as we'll be consuming all the events from Kafka later. Furthermore this migration duplicates the events table and keeps the old table as a backup so we can always restore it.
```

> No. During the migration we take advantage of Kafka storing all the incoming events during the brief period when ingestion from Kafka to ClickHouse is stopped. while we write the last partition to the new table and swap the tables.

Reviewer: lingered?
```suggestion
```

Author: oops

### Will this migration stop ingestion?

> Yes, note that:

Reviewer: ```suggestion
Yes, but please note that:
```

1. There will *not* be any data loss as we'll be consuming all the events from Kafka later.
1. Ingestion is stopped only for a brief period of time, when we are processing the last partition of the old events table and renaming the tables.
1. Ingestion will only be stopped from Kafka to ClickHouse, which is the last step in the ingestion pipeline (see [architecture](https://posthog.com/docs/self-host/architecture)).

### Will I see inconsistent data during the migration?

> Yes, for the brief period of time. When events ingestion from Kafka to ClickHouse is stopped we are still ingesting persons data. For example we might see a person property that was just set, but we wouldn't see the event that set it yet. Once the migration (or rollback) has finished and we caught up everything will be consistent again.

Reviewer: ```suggestion
Yes, for the brief period of time. When event ingestion from Kafka to ClickHouse is stopped we will still process person data. For example we might see a person property that was just set, but we wouldn't see the event that set it yet. Once the migration (or rollback) has finished and we caught up everything will be consistent again.
```
